stonesoup/detector/tensorflow.py

```diff
@@ -0,0 +1,198 @@
+        : set of :class:`~.Detection`
+            Detections generated in the time step. The detection state
+            vector is of the form ``(x, y, w, h)``, where ``x, y`` denote
+            the relative coordinates of the top-left corner of the bounding
+            box containing the object, while ``w, h`` denote the relative
+            width and height of the bounding box. Additionally, each detection
+            carries the following meta-data fields:
+
+            - ``raw_box``: The raw bounding box, as generated by TensorFlow.
+
+            - ``class``: A dict with keys ``id`` and ``name`` relating to the \
+            id and name of the detection class, as specified in the label map.
+
+            - ``score``: A float in the range ``(0, 1]`` indicating the \
+            detector's confidence
+        """
+
+        if self.run_async:
+            yield from self._detections_gen_async()
+        else:
+            yield from self._detections_gen()
+
+    def _capture(self):
+        for timestamp, frame in self.sensor:
+            self._thread_lock.acquire()
+            self._buffer = frame
+            self._thread_lock.release()
+
+    def _detections_gen(self):
+        for timestamp, frame in self.sensor:
+            detections = self._get_detections_from_frame(frame)
+            yield timestamp, detections
+
+    def _detections_gen_async(self):
+        while self._capture_thread.is_alive():
+            if self._buffer is not None:
+                self._thread_lock.acquire()
+                frame = copy(self._buffer)
+                self._buffer = None
+                self._thread_lock.release()
+
+                detections = self._get_detections_from_frame(frame)
+
+                yield frame.timestamp, detections
+
+    def _get_detections_from_frame(self, frame):
+        # Expand dimensions since the model expects images
+        # to have shape: [1, None, None, 3]
+        image_np_expanded = np.expand_dims(frame.pixels, axis=0)
+        # Extract image tensor
+        image_tensor = self._graph.get_tensor_by_name(
+            'image_tensor:0')
+        # Extract detection boxes
+        boxes = self._graph.get_tensor_by_name(
+            'detection_boxes:0')
+        # Extract detection scores
+        scores = self._graph.get_tensor_by_name(
+            'detection_scores:0')
+        # Extract detection classes
+        classes = self._graph.get_tensor_by_name(
+            'detection_classes:0')
+        # Extract number of detections
+        num_detections = self._graph.get_tensor_by_name(
+            'num_detections:0')
+
+        # Perform detection
+        (boxes, scores, classes, num_detections) = self._sess.run(
+            [boxes, scores, classes, num_detections],
+            feed_dict={image_tensor: image_np_expanded})
+
+        classes = np.squeeze(classes).astype(np.int32)
+        boxes = np.squeeze(boxes)
+        scores = np.squeeze(scores)
```

Comment:
I was getting lots of detections with box of `[0, 0, 0, 0]`. Used the below to solve it:
```diff
diff --git a/stonesoup/detector/tensorflow.py b/stonesoup/detector/tensorflow.py
index 6c5f2db..28aef7c 100644
--- a/stonesoup/detector/tensorflow.py
+++ b/stonesoup/detector/tensorflow.py
@@ -171,9 +171,10 @@ class TensorFlowBoxObjectDetector(Detector):
             [boxes, scores, classes, num_detections],
             feed_dict={image_tensor: image_np_expanded})
 
-        classes = np.squeeze(classes).astype(np.int32)
-        boxes = np.squeeze(boxes)
-        scores = np.squeeze(scores)
+        num_detections = int(num_detections)
+        classes = classes[0, :num_detections].astype(np.int32)
+        boxes = boxes[0, :num_detections]
+        scores = scores[0, :num_detections]
 
         # Empty detection set
         frame_height, frame_width, _ = frame.pixels.shape
```

Could keep _squeeze_ , but thought as I was slicing anyway I could just slice it in one.

Reply:
Well spotted. Should be sorted now.